Alembic plugin

Narann opened the pull request:
I open this PR to start preliminary discussion about alembic support. As I'm not fluent in cpp, there is certainly many points where the code is badly organized.

This code is WIP and not intends to be merged yet but as we have xform and dform support, I think it can be interesting to start this discussion.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2017 Esteban Tovagliari, The appleseedhq Organization
```
Put your name in here!

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+{
+
+// roll into given matrix stack and return the flatten one
+asf::Matrix4d flatten_xform(std::vector<asf::Matrix4d> mtx_stack)
```
Pass by const ref to avoid an expensive copy.

dictoon replied:
Indent contents of anonymous namespaces.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+    auto out_mtx = asf::Matrix4d::identity();
+
+    // mutiply the each matrix of the stack from bottom to top
+    for(auto mtx : mtx_stack)
```
Always insert a space after `if`, `for`, `while`, etc.

Coding & formatting guidelines: https://github.com/appleseedhq/appleseed/wiki/Coding-Philosophy-and-Guidelines

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+// roll into given xform sequence stack and return the flatten one
+asr::TransformSequence flatten_xform_seq(const std::vector<asr::TransformSequence>& xform_seq_stack)
+{
+    auto out_xform_seq = asr::TransformSequence();
```
I think this creates a copy.

Simply do this:

    asr::TransformSequence out_xform_seq;

est77 replied:
I suggest moving the transform stack into its own class.
You can optimize the code if instead of pushing a transform and then looping to flatten it everytime is needed,
you push matrix * stack.top(). Then you don't need to loop to flatten.
You can also simplify it by considering a simple matrix as a transform sequence with only one sample and unify the two cases.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+    if(xform_seq_stack.size() == 1)
+    {
+        // we don't need to flatten anything
+        // TODO: Maybe we can return here?
```
Yes, I think we should.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+// return if given time samples are linearly time spaced
+bool linearly_sampled(const Alembic::Abc::chrono_t start_time,
+                      const Alembic::Abc::chrono_t end_time,
+                      std::vector<Alembic::Abc::chrono_t> samples)
```
Pass vector by const ref.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+}
+
+// return if given time samples are linearly time spaced
+bool linearly_sampled(const Alembic::Abc::chrono_t start_time,
```
Add a verb to the method's name: `is_linearly_sampled`

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+    for(const auto t : samples)
+    {
+        std::cout << t << " " << t-t_accum << " " << (std::fabs(t-t_accum) > epsilon) << std::endl;
+        if (std::fabs(t-t_accum) > epsilon)
```
We have more robust ways to compare floating point values:

To check if `x` and `y` are approximately equal:

    foundation::feq(x, y)

Same thing with a custom epsilon value:

    foundation::feq(x, y, 1.0e-6)

To check if `x` is approximately zero:

    foundation::fz(x)

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+
+            for(auto i = 0; i < pos_count; i++)
+            {
+                const auto v = asr::GVector3(pos[i][0], pos[i][1], pos[i][2]);
```
It's really best not to abuse `auto` here, and do this instead:

     const asr::GVector3 v(pos[i][0], pos[i][1], pos[i][2]);

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+
+            const auto sample = schema.getValue(sample_sel);
+
+            const auto pos = sample.getPositions()->get();
```
What does `get()` return? Is this making an expensive copy? It's impossible to tell... I'd suggest not using `auto` when the type is not obvious.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,833 @@
+                const auto pt1 = face_indices[f_i++];
+                const auto pt2 = face_indices[f_i++];
+
+                auto tri0 = asr::Triangle(pt0, pt2, pt1);
```
asr::Triangle tri0(pt0, pt2, pt1);

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.h:
```diff
@@ -0,0 +1,148 @@
+        const asr::ParamArray&          params);
+
+    // Return a string identifying the model of this entity.
+    const char* get_model() const override
```
Implement all methods in the .cpp file.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.h:
```diff
@@ -0,0 +1,148 @@
+
+    void foo(const asr::Assembly* assembly, Alembic::Abc::IObject o);
+
+    static const char* Model;
```
You don't need that. Just declare a _single_ `Model` constant in an anonymous namespace in the cpp file, and use it in all `get_model()` methods.

Narann commented:
Thanks for the review! It actually provide me the obvious things I'm not doing. I will do a global pass on every points.

dictoon commented:
❤❤❤

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+
+    // extract an appleseed xform sequence from given alembic xform schema
+    asr::TransformSequence extract_xform_seq(
+            const AbcGeom::IXformSchema& schema,
```
Fix indentation.
Align parameters.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+    //                  (floor)  ,<-'              '-->,  (ceil)
+    // returned samples    ------|-----------|---------|--------->
+    std::vector<AbcIndex_t> schema_to_sample_times(
+        const float shutter_open,
```
Align parameters.

Narann replied:
You mean from this:
```c++
    std::vector<AbcIndex_t> schema_to_sample_times(
        const float shutter_open,
        const float shutter_close,
        const size_t num_samples,
        const Abc::TimeSamplingPtr time_sampling)
```
To this:
```c++
    std::vector<AbcIndex_t> schema_to_sample_times(
        const float                shutter_open,
        const float                shutter_close,
        const size_t               num_samples,
        const Abc::TimeSamplingPtr time_sampling)
```
?

dictoon replied:
Exactly!

Narann replied:
I did it, Github diff just didn't catch it.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+        }
+        else  // more than 1 sample
+        {
+            // get floor and ceil sample indices to don't miss any
```
to _not_ miss any

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+
+namespace
+{
+    // shortcuts to save space
```
Full line comments should start with a capital and end with a period:

    // Shortcuts to save space.

(Fix everywhere.)

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+    asf::Matrix4d flatten_xform(const std::vector<asf::Matrix4d>& mtx_stack)
+    {
+        // the matrix we will return
+        asf::Matrix4d out_mtx = asf::Matrix4d::identity();
```
A small optimization would be to initialize `out_mtx` with the first matrix from the stack (but then you can't use a range-based `for` loop).

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+    // roll into given xform sequence stack and return the flatten one
+    asr::TransformSequence flatten_xform_seq(const std::vector<asr::TransformSequence>& xform_seq_stack)
+    {
+        assert(xform_seq_stack.size() > 0);
```
Prefer `empty()`:

    assert(!xform_seq_stack.empty());

Narann replied:
What about ` assert(xform_seq_stack.size())` so there is no negative operator (less cognitive load).

dictoon replied:
That has the wrong semantic: `size()` does not return a boolean, it doesn't make sense to check if `size()` is true or false. The fact that 0 evaluates to false and other values to true is inconsequential.

dictoon replied:
Side remark: In C#, `System.Linq` provides `Any()`:

    Debug.Assert(xform_seq_stack.Any());

Narann replied:
Side remark: In Python:

```python
if xform_seq_stack:
```
B-[

dictoon replied:
Even

    assert xform_seq_stack

:) But again not quite happy about the semantics it carries (or the lack thereof)... Maybe it's just me.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+            // xform sequence we will return
+            asr::TransformSequence out_xform_seq;
+
+            for (const asr::TransformSequence& xform_seq : xform_seq_stack)
```
That's a place where `auto` would make sense.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+
+    // return if given time samples are linearly time spaced between start_time
+    // and end_time.
+    const bool are_linearly_sampled(const AbcChrono_t start_time,
```
Format like you did with other methods (all parameters go on their own line).

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+        // get each sample time and compare with a linearly sampled time
+        for (const AbcChrono_t t : samples)
+        {
+            if (asf::feq(t, t_accum))
```
Are you sure about the logic here? Shouldn't it be `!asf::feq()`?

Narann replied:
Done.
Good catch! This code is not used yet as its purpose was to prevent the user of bad sampling. I just don't know yet _when_ I will check and report.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+                return false;
+            }
+
+            t_accum += increm;
```
In theory errors could accumulate here. It would be more robust (albeit slightly slower) to recompute a linear sampled `t` value at each iteration, for instance using `foundation::fit()`.

Narann replied:
Maybe I'm tired but I don't see how a remap can help here.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+        else  // more than 1 sample
+        {
+            // get floor and ceil sample indices to don't miss any
+            const std::pair<AbcIndex_t, AbcChrono_t>& open_pair =
```
Just checking: are you sure that `getFloorIndex()` returns a _reference_ and not a value?

Narann replied:
No it doesn't. I was doing this to save a copy. Should I remove the ref identifier?

dictoon replied:
> Should I remove the ref identifier?

Yes, because as the code stands right now, you're keeping a reference to a temporary object that no longer exists outside `getFloorIndex()`, i.e. it's broken (even if it may occasionally work).

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+            const std::pair<AbcIndex_t, AbcChrono_t>& close_pair =
+                time_sampling->getCeilIndex(shutter_close, num_samples);
+
+            sample_indices.reserve(close_pair.first-open_pair.first+1);
```
Add spaces around `+` operator.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+            const auto mtx = asf::Matrix4d(abc_mtx);
+            const auto xform = asf::Transformd::from_local_to_parent(mtx);
+
+            // as all samples are time offsetted, we offset them back
```
*offsetted* -> *offset*

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+    void log_archive(Abc::IArchive& archive,
+                     AbcCoreFac::IFactory::CoreType core_type)
+    {
+        RENDERER_LOG_INFO("archive name: %s\n"
```
Move all arguments to their own line.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+            archive.getName().c_str(),
+            archive.getNumTimeSamplings(),
+            archive.getArchiveVersion());
+        switch(core_type)
```
Add space after `switch`.

Narann replied:
Done (here and everywhere else).

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+                RENDERER_LOG_WARNING("        core type: Unknown");
+                break;
+            default:
+                break;  // this should never happen (tm)
```
In that case, assert:

    assert(!"This should never happen.");

Or use `assert_otherwise` (defined in `foundation/utility/otherwise.h`).

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+    void log_obj(const Abc::IObject& obj)
+    {
+        RENDERER_LOG_INFO("%s\n"
+                          "  children count: %zd",
```
I'm not sure what's the type returned by `getNumChildren()`, but if it's `std::size_t` then use `FMT_SIZE_T` defined in `foundation/platform/types.h`.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+
+        if (metadata.size())
+        {
+            for (const std::pair<std::string, std::string>& it : metadata)
```
Ideal case for `const auto&`.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+
+            if (children_count)
+            {
+                for (size_t i = 0; i < root.getNumChildren(); i++)
```
The following idiom guarantees that the upper bound is not reevaluated at each iteration:

    for (size_t i = 0, e = root.getNumChildren(); i < e; i++)

Narann replied:
Done.

Nice trick!

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+                if (m_verbose)
+                    log_polymesh_schema(schema);
+
+                /*const auto sample_times = schema_to_sample_timesOO(m_shutter_open+m_time_offset,
```
Remove dead code.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+                    switch(n_param.getScope())
+                    {
+                        case AbcGeom::kConstantScope:
+                            std::cout << "normal kConstantScope" << std::endl;
```
Remove debug code.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+                asf::Dictionary()
+                    .insert("name", Model)
+                    .insert("label", "Alembic Camera")
+                    .insert("default_model", "true");
```
The default camera model is the pinhole camera; remove `default_model` altogether.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+        }
+
+      private:
+
```
Remove blank line.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+      private:
+
+        const char* Model = "alembic_assembly";
+
```
Remove blank line.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+        }
+
+      private:
+
```
Remove blank line.

Narann replied:
Done.

dictoon commented on sandbox/samples/cpp/alembicassembly/alembicassembly.cpp:
```diff
@@ -0,0 +1,1525 @@
+
+            if (xform_seq_stack.size())
+            {
+                const asr::TransformSequence xform_seq = flatten_xform_seq(xform_seq_stack);
```
Get rid of temporary variable?

Narann replied:
Done.

I like to have one line doing one thing and let the compiler optimize. But in this case it's pretty obvious.

Narann commented:
Thanks for the kind words, I will fix those asap.

Narann commented:
I think it can be merged now. Rougth instruction:

You need to compile alembic (I did in .so).

Then:
```bash
# to make
export PATH=<alembic_dir>/bin:$PATH
export LD_LIBRARY_PATH=<alembic_dir>/lib:$LD_LIBRARY_PATH
cd <appleseed>/sandbox/samples/cpp/alembicassembly/build
make
```
Then:
```bash
# to render
export PATH=<alembic_dir>/bin:$PATH
export LD_LIBRARY_PATH=<alembic_dir>/lib:$LD_LIBRARY_PATH
cd <appleseed>/sandbox/bin/Release
./appleseed.cli <appleseed>/sandbox/samples/cpp/alembicassembly/cam.appleseed -o <appleseed>/sandbox/samples/cpp/alembicassembly/foo.png
```
The last line can be .studio instead of .cli.

There is a simple scene samples but you would need to provide your own alembic to test. We should do some alembic tests at some points.

This assembly provide both assembly and camera plugin. Documentation is here:

https://github.com/appleseedhq/appleseed/wiki/Alembic-Assembly
https://github.com/appleseedhq/appleseed/wiki/Alembic-Camera

Don't hesitate if you have question but for me, it's ready to go.

dictoon commented:
Amazing Dorian, thanks for the great work!

I haven't checked the code yet. If I review it and make comments, will you have time to address them? Or should we "adopt" the code and take care of the rest ourselves? Either way is fine, just le me know.

Narann commented:
Thanks. You already reviewed the code month ago. I did all modifications and the branch hasn't change that much.

There is one single comment to address (ref usage where I should not). I will do that quickly.

If the whole design is OK, I now have two option:
1) I delete addressed comments to keep this PR clean so you can do more comments and I can address them.
2) Merge then others modify things.

If the design is bad, yes, we should continue to discuss this but as we kinda discussed it at our last meetup, then I inspired myself from other exporters (in the Alembic repo) then you reviewed the code a first type, I think the potential design issues would have raised sooner.

I think this maybe time to let "real" dev do the job now. I suggest to merge, don't worry too much "my" code. Yes, it's "my baby" but I think you guys will take best decisions for it.

If the questions are only related to "you should(n't) use ref here", "align parameters", it will be quicker if you take care of it (I think).

I realize there is a lack not in the code itself but in the test suite. Alembic is too broad to have simple tests, you have to have complex alembics to test.

I think I will focus on writing a script that generate various alembic situation so we can have a "rebuildable" test suite for our alembic support. This script would be runnable in Maya to generate many different alembics.

I think this effort is something not that many appleseed devs can take care easily (while keeping care of my code is something you can do).

With such test suite, we should be able to have regression test for our alembic support.

What do you think ?

dictoon commented:
That sounds excellent! I will merge then.

Out of curiosity, why is a script generating a test suite better than a bunch of Alembic files?

Narann commented:
Thanks!

I think having a script generating alembic files is better to explicitly states what's inside alembic for everyone (it's written in the script, no guess). This is something I realized during my tests. Depicts knowing what was supposed to be in the alembic files, I often open them again in Maya to compare/be sure and often wanted "this one with this export option". The idea is to have functions generating scenes, and export presets to export with specific options (different sampling, color set, uv set, etc).

I already have something an I'm working to also generate a subsampled playblast at export time (the maya scene to export) and also subsampled playblast of the alembic file (no interpolation between samples). I also plan to have jpg file of every frame, every sample, etc. And generate a world position value file for every object at every sample, same for vertex 0. So no guess. With that, you can see what the export is supposed to be on the Maya side, on the alembic side, with picture and everything without having Maya. And have values to compare.

dictoon commented:
Thans for the explanation, makes a lot of sense!
